compiler/env/mem/OMRSegmentAllocator.cpp

```diff
@@ -0,0 +1,257 @@
+
+   // first, see if we need to allocate a new chunk of backing memory
+   size_t remainingSpace = _currentBackingMemoryChunk.get().remaining();
+   if (remainingSpace < roundedSize && _backingMemoryAllocator.fitsInMinimumSizeSegment(roundedSize))
+      {
+      MEMLOG(2, "current chunk only %lu remaining, need to allocate a new one\n", remainingSpace);
+
+      // doesn't fit in current chunk, but it would fit in a fresh chunk
+      // first split remaining space in current chunk into small segments and put onto free list
+      claimFreeSegments(_currentBackingMemoryChunk.get());
+      _currentBackingMemoryChunk = TR::ref(_backingMemoryAllocator.allocate(_allocationBlockSize));
+
+      // update remainingSpace so allocation will come from the (new) current chunk
+      remainingSpace = _currentBackingMemoryChunk.get().remaining();
+      MEMLOG(2, "new chunk %p allocated, remaining now %lu\n", &(_currentBackingMemoryChunk.get()), remainingSpace);
+      }
+
+   // allocate from current chunk if it will fit
+   if (remainingSpace >= roundedSize)
+      {
+      MEMLOG0(2, "allocation will come from current chunk\n");
+      return allocateFromSegment(roundedSize, _currentBackingMemoryChunk.get());
+      }
+
+   // alas, we need a dedicated backing memory allocation for this allocation request
+   TR::MemorySegment &newSegment = _backingMemoryAllocator.allocate(roundedSize);
+   MEMLOG(2, "need dedicated segment, allocated %p size %lu\n", &newSegment, newSegment.size());
+   _allocatedSegments.push_front( TR::ref(newSegment) );
+   _bytesAllocated += roundedSize;
+   return newSegment;
+   }
+
+void
+OMR::SegmentAllocator::deallocate(TR::MemorySegment & segment) throw()
+   {
+   MEMLOG(1, "deallocate %p with size %lu request\n", &segment, segment.size());
+   size_t const segmentSize = segment.size();
+   _bytesAllocated -= segmentSize;
+   segment.reset(); // claim the entire segment
+
+   // segment should no longer be consider "allocated"
+   bool found=false;
+   auto segmentToRemove = _allocatedSegments.begin();
+   while (segmentToRemove != _allocatedSegments.end())
+      {
+      auto ref = *segmentToRemove;
+      MEMLOG(3, "_allocatedSegments iterating segment %p\n", &(ref.get()));
+      if (segment == ref.get())
+         {
+         found=true;
+         break;
+         }
+      segmentToRemove++;
+      }
+   TR_ASSERT(found, "deallocate() request for segment not in list of allocated segments");
+   // we cannot erase segmentToRemove from _allocatedSegments yet because _allocatedSegments owns the TR::MemorySegment object
+   //  so if we were to erase it now, that memory would be reclaimed before we're done with the freeing operations below
+
+   auto ref = *segmentToRemove;
+   MEMLOG(2, "deallocate %p ( %p ) was found in allocatedSegments\n", &segment, &(ref.get()));
+
+   // small segments that match the block size can just be put onto free list
+   if (segmentSize == _allocationBlockSize)
```

Comment:
This is a pedantic comment,so ignore at will :): isn't this just a special case of `else if (_backingMemoryAllocator.fitsInMinimumSizeSegment(segmentSize))`?

Reply:
well, remember that the backing memory allocator and the segment allocator don't need to use the same segment allocation size. And the segment allocator can satisfy a request with a segment that's a multple of the segment allocator's default size but still less than the backing memory allocator's default size.

For example, the backing memory allocator may be allocating no smaller than 1MB segments and the segment allocator may be allocating no smaller than 64KB segments. The segment allocator can handle a memory request for 256KB (to satisfy a request that's any larger than 192KB) from the current backing memory chunk. When that segment comes back, it can't just go onto the free list: it needs to be carved up into 64KB segments first.

Reply:
Ah right; heh that's a subtlety I missed :)